### ISSUE TITLE
[Discussion] Display categorised repository history

### ISSUE THREAD

**Issue by RussKie** [gitextensions/gitextensions#4878]
I am currently reworking how we manage user's repository history - recent and favourites.
As part of that rework to provide consistent UX in terms of having access to favourites I am seeking input for how to best present categorised repositories on the main screen.

There are few options:

1. Do not group, sort according to the user's preferences (A-Z or date accessed). 
_Pros:_ no hidden entries
_Cons:_ the drop down can be very long
It will look something like this:
![image](https://user-images.githubusercontent.com/4403806/39243067-9ddf5b52-48cf-11e8-9b38-a24fdbf571c5.png)

2. Group by category **after** recent, sort according to the user's preferences (A-Z or date accessed). 
_Pros:_ recent are more readily accessible
_Cons:_ if user has too many recents, favourites may be off the screen and will need to be scrolled to
It will look something like this:
![image](https://user-images.githubusercontent.com/4403806/39244424-f8ce6040-48d3-11e8-9c23-b9902690cf61.png)

3. Group by category **before** recent, sort according to the user's preferences (A-Z or date accessed). 
_Pros:_ favourites are more readily accessible
_Cons:_ if user has too many favourites, recents may need to be scrolled to
It will look something like this:
![image](https://user-images.githubusercontent.com/4403806/39244320-b32f07f6-48d3-11e8-8b67-3cbbe01069cb.png)

Thoughts?

Relates to #3693

**Comment by RussKie** [comment]
Thank you
Few comments:

> display max 7 items of each (because the brain can't handle more without efforts) and add menus items "Other favorites..." and "Other recents..." to display all the others (and hidden if there is no item inside).

if a user has multiple categories, how do we pick top 7?
e.g. @francoas has [hundreds of favourites spread across dozes of categories](https://github.com/gitextensions/gitextensions/pull/3693#issuecomment-310126674).

> I'm not very fond of displaying categories here because categories are overkill for most of users (Or perhaps displayed in a sub item "Categories"?). If the user need it, it will close the repository to go to the dashboard. Here is a quick access...

If you don't have favourites, they won't appear in the list.
Forcing users to go to the dashboard to open another repo is a poor UX, this dropdown is a perfect shortcut.

> List is automaticaly purged when some repositories are no more existing (only on drives existing, to not remove the ones on removable devices)

I'm not sure it is easily achievable. A user may as well map drive D:\ and then at some point it may become unavailable. How do we know it is safe to remove entries pointed to D: drive?
Also you can mount a folder via [junctions](https://en.wikipedia.org/wiki/NTFS_junction_point)...

**Comment by RussKie** [comment]
Another possibility is to have a top level menu item "Favourites" (or "Categories" or whatever) and then it would show multiple entries for each category

```
|- Favourites
|   | - Git Extensions
|   |   | - d:\bla\Git Extensions.git
|   | - Test
|   |   | - d:\bla\test1.git
|   |   | - d:\bla\test2.git
|   |   | - d:\bla\test3.git
|   | - 3rd party
|- c:\foo\bar.git  
|- c:\foo\bar2.git  
``

**Comment by pmiossec** [comment]
>e.g. @francoas has hundreds of favourites spread across dozes of categories.

I know that but let me believe that wanting to display more than 100 items in a drowdown list is a poor UX.
I think that you try to solve a case that very few users have. And for those, a search/filter feature is better. or use the dashboard where a lot more data could be displayed (without navigating between sub menus which are quite always painfull...) 
I think that we should focus on the case of more than 95% of users that have less than 20 repositories (either favorites or recent)

What I imagine:

```
|- Favourite Repo 1
|- Favourite Repo 2
|- Favourite Repo 3
|- Favourite Repo 4
|- Favourite Repo 5
|- Other Favourites....
|   | - Favourite Repo 6
|   | - ......
|   | - Favourite Repo n
|----------------------
|- Recent Repo 1
|- Recent Repo 2
|- Recent Repo 3
|- Recent Repo 4
|- Recent Repo 5
|- Other Recents....
|   | - Recent Repo 6
|   | - ......
|   | - Recent Repo n
|----------------------
|- Categories
|   | - Git Extensions
|   |   | - d:\bla\Git Extensions.git
|   | - Test
|   |   | - d:\bla\test1.git
|   |   | - d:\bla\test2.git
|   |   | - d:\bla\test3.git
|   | - 3rd party
```

**Comment by RussKie** [comment]
Categories and favourites are the same thing

**Comment by jbialobr** [comment]
I like the option 3 the most - the categories submenus keep position even though new non categorized repos are added. I think we can start with any option and leave it for further improvements (customizations, as I guess there will be many mutually exclusive opinions here) in separate PRs.

**Comment by francoas** [comment]
Since I have been quoted many times, here is my humble opinion

But first of all: even I have "hundreds of repos in many categories", it is true that I often work with the same bunch during a certain period.
So I would take advantage of the "recents" list 70% of the time and use the categories 30% of the time (when I switch context or I am asked for a maintainance...)

That said, I like options 2 with 10 recents at most (like visual studio does, no need for "other recents") and then JUST a top menu voice "categories" that can be expanded in sub menus and can be configured.

>Another possibility is to have a top level menu item "Favourites" (or "Categories" or whatever) and then it would show multiple entries for each category

Exactly

But actually, there is no need for categories in that menu, provided there is a voice "Categories..." (or favorite... I agree they are the same) in the Start menu just after the "Recent Repositories..." voice

So, to summarize:

A. In the Start menu I would add the "Categories..." (or Favourites) voice just after (or before) the "Recents Repositories..." voice.

B: in the drop down I would use option 2 with a top level menu item for categories, or even keep just the recents (as it is now), provided that A is true

Suggestion: I would use the icons just for the top menu voices ("recent" and "categories") not for each repos.

 @RussKie: you are doing a great job

**Comment by RussKie** [comment]
> That said, I like options 2 with 10 recents at most (like visual studio does, no need for "other recents") 

This is a configurable setting, default is 30.
![image](https://user-images.githubusercontent.com/4403806/39406193-e0a1a994-4bf5-11e8-858d-631ec7d9c757.png)

**Comment by RussKie** [comment]
Thank you all, based on the feedback it will look similar the following:

![image](https://user-images.githubusercontent.com/4403806/39406240-d4f38fda-4bf6-11e8-9ddf-60f97ed9646b.png)

![image](https://user-images.githubusercontent.com/4403806/39406753-03a1d53c-4bff-11e8-9c53-34e33e523fcf.png)
